<a href="https://colab.researc.google.com/github/manmeetsingh089/sentimental-analysis-of-twitter-movie-review/blob/main/DMT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:

from google.colab import drive

drive.mount('/content/gdrive/', force_remount=True)

Mounted at /content/gdrive/


# **Libraries**

---



In [ ]:
import pandas as pd 
import csv
import numpy as np
import re
import nltk 
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer 
from nltk import tokenize
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Input, SimpleRNN, Embedding, Dropout, Activation, Flatten, Bidirectional, GlobalMaxPool1D,GRU
from tensorflow.keras.models import Model, Sequential
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Conv1D, MaxPooling1D, Flatten, Dense,Dropout
import pandas as pd

# **Pre**-Processing

---



In [ ]:
df = pd.read_csv(r"IMDB Dataset.csv",engine='python',error_bad_lines=False,)
df.head()

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [ ]:
df.shape

(49999, 2)

In [ ]:
label = LabelEncoder()
df['sentiment']=label.fit_transform(df['sentiment'])

In [ ]:
df.groupby(df["sentiment"]).count()

review
sentiment        
0           24999
1           25000

In [ ]:
df['length'] =df['review'].apply(lambda x: len(x))
df.head()

review  sentiment  length
0  One of the other reviewers has mentioned that ...          1    1761
1  A wonderful little production. <br /><br />The...          1     998
2  I thought this was a wonderful way to spend ti...          1     926
3  Basically there's a family where a little boy ...          0     748
4  Petter Mattei's "Love in the Time of Money" is...          1    1317

In [ ]:
df.head()

review  sentiment  length
0  One of the other reviewers has mentioned that ...          1    1761
1  A wonderful little production. <br /><br />The...          1     998
2  I thought this was a wonderful way to spend ti...          1     926
3  Basically there's a family where a little boy ...          0     748
4  Petter Mattei's "Love in the Time of Money" is...          1    1317

In [ ]:
vocab_size = 10000
maxlen = 6000  
embedding_dim = 32
epochs = 50
batch_size = 100
trunc_type='post'
oov_tok = "<OOV>"

In [ ]:
t=df['review']
y=df['sentiment']

In [ ]:
x.head()

0    One of the other reviewers has mentioned that ...
1    A wonderful little production. <br /><br />The...
2    I thought this was a wonderful way to spend ti...
3    Basically there's a family where a little boy ...
4    Petter Mattei's "Love in the Time of Money" is...
Name: review, dtype: object

In [ ]:
ps = PorterStemmer()
corpus=[]
x=[]
for i in range(len(t)):
  review=re.sub('http://\S+|https://\S+', '', t[i])
  review=re.sub("[^a-zA-z]"," ",review)
  review=re.sub('<.*?>','',review)
  review =re.sub(r'[^\w\s]', '', review)
  review=review.lower()
  review = review.split()
  review = [ps.stem(word) for word in review if word not in set(stopwords.words("english"))]
  review= " ".join(review)
  x.append(review)
  corpus.append(review)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


# **Tokenization of reviews**

---



In [ ]:
tokenizer = Tokenizer(num_words=vocab_size, oov_token=oov_tok)
tokenizer.fit_on_texts(x)

In [ ]:
X = tokenizer.texts_to_sequences(x)
X = pad_sequences(X, maxlen = maxlen, truncating=trunc_type)
y = np.asarray(y)

print("Number of Samples:", len(X))
print(X[0])
print("Number of Labels: ", len(y))
print(y[0])

Number of Samples: 49999
[   0    0    0 ...  128 4059  481]
Number of Labels:  49999
1


# **Training and Tesing Split**

---



In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=101, test_size= 0.2) 

In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((39999, 6000), (10000, 6000), (39999,), (10000,))

In [ ]:
X_train[0]

array([   0,    0,    0, ...,    5,    2, 2839], dtype=int32)

# **Model CNN+RNN**

---



In [ ]:
model = Sequential()
model.add(Embedding(vocab_size, embedding_dim, input_length=maxlen))
model.add(Conv1D(32, kernel_size=3,padding='same', activation='relu'))
model.add(MaxPooling1D(pool_size=2))
model.add(Dropout(0.25))
model.add(SimpleRNN(64))
model.add(Flatten())
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
print(model.summary())

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_9 (Embedding)     (None, 6000, 32)          320000    
                                                                 
 conv1d_9 (Conv1D)           (None, 6000, 32)          3104      
                                                                 
 max_pooling1d_9 (MaxPooling  (None, 3000, 32)         0         
 1D)                                                             
                                                                 
 dropout_9 (Dropout)         (None, 3000, 32)          0         
                                                                 
 simple_rnn_2 (SimpleRNN)    (None, 64)                6208      
                                                                 
 flatten_8 (Flatten)         (None, 64)                0         
                                                      

In [ ]:
model.fit(X_train, y_train, epochs=5,batch_size=256,verbose = 1)

Epoch 1/5
157/157 [==============================] - 656s 4s/step - loss: 0.6598 - accuracy: 0.5905
Epoch 2/5
157/157 [==============================] - 648s 4s/step - loss: 0.3519 - accuracy: 0.8512
Epoch 3/5
157/157 [==============================] - 649s 4s/step - loss: 0.4113 - accuracy: 0.8110
Epoch 4/5
157/157 [==============================] - 655s 4s/step - loss: 0.7801 - accuracy: 0.7135
Epoch 5/5
157/157 [==============================] - 645s 4s/step - loss: 0.5526 - accuracy: 0.7224


In [ ]:
test_acc=model.evaluate(X_test,y_test)

313/313 [==============================] - 83s 263ms/step - loss: 0.5599 - accuracy: 0.7043


In [ ]:
print('accuracy for CNN+RNN model is %.03f'%(test_acc[1]*100))

accuracy for CNN+RNN model is 70.430


# **Accuracy for CNN+RNN model is 70.430**

# **Model CNN+GRU**

In [ ]:
model1 = Sequential()
model1.add(Embedding(vocab_size, embedding_dim, input_length=maxlen))
model1.add(Conv1D(32, kernel_size=3,padding='same', activation='relu'))
model1.add(MaxPooling1D(pool_size=2))
model1.add(Dropout(0.25))
model1.add(GRU(64))
model1.add(Flatten())
model1.add(Dense(1, activation='sigmoid'))
model1.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
print(model1.summary())

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_5 (Embedding)     (None, 6000, 32)          320000    
                                                                 
 conv1d_5 (Conv1D)           (None, 6000, 32)          3104      
                                                                 
 max_pooling1d_5 (MaxPooling  (None, 3000, 32)         0         
 1D)                                                             
                                                                 
 dropout_5 (Dropout)         (None, 3000, 32)          0         
                                                                 
 gru_4 (GRU)                 (None, 64)                18816     
                                                                 
 flatten_5 (Flatten)         (None, 64)                0         
                                                      

In [ ]:
model1.fit(X_train, y_train, epochs=5,steps_per_epoch=50 ,batch_size=256,verbose = 1)

Epoch 1/5
50/50 [==============================] - 551s 9s/step - loss: 0.6864 - accuracy: 0.5620
Epoch 2/5
50/50 [==============================] - 458s 9s/step - loss: 0.4607 - accuracy: 0.7819
Epoch 3/5
50/50 [==============================] - 468s 9s/step - loss: 0.3450 - accuracy: 0.8536
Epoch 4/5
50/50 [==============================] - 464s 9s/step - loss: 0.2652 - accuracy: 0.8929
Epoch 5/5
50/50 [==============================] - 460s 9s/step - loss: 0.2505 - accuracy: 0.8995


In [ ]:
test_acc1 = model1.evaluate(X_test,y_test)

313/313 [==============================] - 121s 383ms/step - loss: 0.2952 - accuracy: 0.8862


In [ ]:
print('accuracy for CNN+GRU model is %.03f'%(test_acc1[1]*100))

accuracy for CNN+GRU model is 88.620


# **Accuracy for CNN+GRU model is 88.620**

# **Model Bert**

---



# **Libraries**

In [ ]:
!pip install -q -U tensorflow-text --quiet

     |████████████████████████████████| 5.9 MB 26.1 MB/s 
     |████████████████████████████████| 578.0 MB 18 kB/s 
     |████████████████████████████████| 1.7 MB 48.6 MB/s 
     |████████████████████████████████| 5.9 MB 57.6 MB/s 
     |████████████████████████████████| 438 kB 109.6 MB/s 


In [ ]:
!pip install -q tf-models-official

     |████████████████████████████████| 2.2 MB 31.2 MB/s 
     |████████████████████████████████| 1.1 MB 72.6 MB/s 
     |████████████████████████████████| 1.3 MB 62.1 MB/s 
     |████████████████████████████████| 238 kB 80.3 MB/s 
     |████████████████████████████████| 43 kB 2.4 MB/s 
     |████████████████████████████████| 38.2 MB 1.5 MB/s 
     |████████████████████████████████| 352 kB 103.7 MB/s 
     |████████████████████████████████| 116 kB 81.6 MB/s 
     |████████████████████████████████| 636 kB 84.2 MB/s 


In [ ]:
import os
import shutil

import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_text as text
from official.nlp import optimization  # to create AdamW optimizer

import matplotlib.pyplot as plt

tf.get_logger().setLevel('ERROR')

# **model**

---



In [ ]:
#@title Choose a BERT model to fine-tune

bert_model_name = 'small_bert/bert_en_uncased_L-2_H-128_A-2'  #@param ["bert_en_uncased_L-12_H-768_A-12", "bert_en_cased_L-12_H-768_A-12", "bert_multi_cased_L-12_H-768_A-12", "small_bert/bert_en_uncased_L-2_H-128_A-2", "small_bert/bert_en_uncased_L-2_H-256_A-4", "small_bert/bert_en_uncased_L-2_H-512_A-8", "small_bert/bert_en_uncased_L-2_H-768_A-12", "small_bert/bert_en_uncased_L-4_H-128_A-2", "small_bert/bert_en_uncased_L-4_H-256_A-4", "small_bert/bert_en_uncased_L-4_H-512_A-8", "small_bert/bert_en_uncased_L-4_H-768_A-12", "small_bert/bert_en_uncased_L-6_H-128_A-2", "small_bert/bert_en_uncased_L-6_H-256_A-4", "small_bert/bert_en_uncased_L-6_H-512_A-8", "small_bert/bert_en_uncased_L-6_H-768_A-12", "small_bert/bert_en_uncased_L-8_H-128_A-2", "small_bert/bert_en_uncased_L-8_H-256_A-4", "small_bert/bert_en_uncased_L-8_H-512_A-8", "small_bert/bert_en_uncased_L-8_H-768_A-12", "small_bert/bert_en_uncased_L-10_H-128_A-2", "small_bert/bert_en_uncased_L-10_H-256_A-4", "small_bert/bert_en_uncased_L-10_H-512_A-8", "small_bert/bert_en_uncased_L-10_H-768_A-12", "small_bert/bert_en_uncased_L-12_H-128_A-2", "small_bert/bert_en_uncased_L-12_H-256_A-4", "small_bert/bert_en_uncased_L-12_H-512_A-8", "small_bert/bert_en_uncased_L-12_H-768_A-12", "albert_en_base", "electra_small", "electra_base", "experts_pubmed", "experts_wiki_books", "talking-heads_base"]

map_name_to_handle = {
    'bert_en_uncased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/3',
    'bert_en_cased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_cased_L-12_H-768_A-12/3',
    'bert_multi_cased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_multi_cased_L-12_H-768_A-12/3',
    'small_bert/bert_en_uncased_L-2_H-128_A-2':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-2_H-128_A-2/1',
    'small_bert/bert_en_uncased_L-2_H-256_A-4':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-2_H-256_A-4/1',
    'small_bert/bert_en_uncased_L-2_H-512_A-8':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-2_H-512_A-8/1',
    'small_bert/bert_en_uncased_L-2_H-768_A-12':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-2_H-768_A-12/1',
    'small_bert/bert_en_uncased_L-4_H-128_A-2':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-4_H-128_A-2/1',
    'small_bert/bert_en_uncased_L-4_H-256_A-4':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-4_H-256_A-4/1',
    'small_bert/bert_en_uncased_L-4_H-512_A-8':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-4_H-512_A-8/1',
    'small_bert/bert_en_uncased_L-4_H-768_A-12':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-4_H-768_A-12/1',
    'small_bert/bert_en_uncased_L-6_H-128_A-2':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-6_H-128_A-2/1',
    'small_bert/bert_en_uncased_L-6_H-256_A-4':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-6_H-256_A-4/1',
    'small_bert/bert_en_uncased_L-6_H-512_A-8':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-6_H-512_A-8/1',
    'small_bert/bert_en_uncased_L-6_H-768_A-12':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-6_H-768_A-12/1',
    'small_bert/bert_en_uncased_L-8_H-128_A-2':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-8_H-128_A-2/1',
    'small_bert/bert_en_uncased_L-8_H-256_A-4':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-8_H-256_A-4/1',
    'small_bert/bert_en_uncased_L-8_H-512_A-8':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-8_H-512_A-8/1',
    'small_bert/bert_en_uncased_L-8_H-768_A-12':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-8_H-768_A-12/1',
    'small_bert/bert_en_uncased_L-10_H-128_A-2':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-10_H-128_A-2/1',
    'small_bert/bert_en_uncased_L-10_H-256_A-4':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-10_H-256_A-4/1',
    'small_bert/bert_en_uncased_L-10_H-512_A-8':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-10_H-512_A-8/1',
    'small_bert/bert_en_uncased_L-10_H-768_A-12':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-10_H-768_A-12/1',
    'small_bert/bert_en_uncased_L-12_H-128_A-2':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-128_A-2/1',
    'small_bert/bert_en_uncased_L-12_H-256_A-4':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-256_A-4/1',
    'small_bert/bert_en_uncased_L-12_H-512_A-8':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-512_A-8/1',
    'small_bert/bert_en_uncased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-768_A-12/1',
    'albert_en_base':
        'https://tfhub.dev/tensorflow/albert_en_base/2',
    'electra_small':
        'https://tfhub.dev/google/electra_small/2',
    'electra_base':
        'https://tfhub.dev/google/electra_base/2',
    'experts_pubmed':
        'https://tfhub.dev/google/experts/bert/pubmed/2',
    'experts_wiki_books':
        'https://tfhub.dev/google/experts/bert/wiki_books/2',
    'talking-heads_base':
        'https://tfhub.dev/tensorflow/talkheads_ggelu_bert_en_base/1',
}

map_model_to_preprocess = {
    'bert_en_uncased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'bert_en_cased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_cased_preprocess/3',
    'small_bert/bert_en_uncased_L-2_H-128_A-2':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-2_H-256_A-4':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-2_H-512_A-8':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-2_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-4_H-128_A-2':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-4_H-256_A-4':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-4_H-512_A-8':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-4_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-6_H-128_A-2':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-6_H-256_A-4':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-6_H-512_A-8':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-6_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-8_H-128_A-2':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-8_H-256_A-4':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-8_H-512_A-8':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-8_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-10_H-128_A-2':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-10_H-256_A-4':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-10_H-512_A-8':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-10_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-12_H-128_A-2':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-12_H-256_A-4':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-12_H-512_A-8':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'small_bert/bert_en_uncased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'bert_multi_cased_L-12_H-768_A-12':
        'https://tfhub.dev/tensorflow/bert_multi_cased_preprocess/3',
    'albert_en_base':
        'https://tfhub.dev/tensorflow/albert_en_preprocess/3',
    'electra_small':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'electra_base':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'experts_pubmed':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'experts_wiki_books':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
    'talking-heads_base':
        'https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3',
}

tfhub_handle_encoder = map_name_to_handle[bert_model_name]
tfhub_handle_preprocess = map_model_to_preprocess[bert_model_name]

print(f'BERT model selected           : {tfhub_handle_encoder}')
print(f'Preprocess model auto-selected: {tfhub_handle_preprocess}')

BERT model selected           : https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-2_H-128_A-2/1
Preprocess model auto-selected: https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3


In [ ]:
bert_preprocess_model = hub.KerasLayer(tfhub_handle_preprocess)

In [ ]:
bert_model = hub.KerasLayer(tfhub_handle_encoder)

In [ ]:
bert_results = bert_model(text_preprocessed)

print(f'Loaded BERT: {tfhub_handle_encoder}')
print(f'Pooled Outputs Shape:{bert_results["pooled_output"].shape}')
print(f'Pooled Outputs Values:{bert_results["pooled_output"][0, :12]}')
print(f'Sequence Outputs Shape:{bert_results["sequence_output"].shape}')
print(f'Sequence Outputs Values:{bert_results["sequence_output"][0, :12]}')

Loaded BERT: https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-2_H-128_A-2/1
Pooled Outputs Shape:(2, 128)
Pooled Outputs Values:[-0.99998695  0.14298673 -0.99944323  0.9977272  -0.9996823   0.9681122
 -0.9964279  -0.9565575   0.09264198 -0.0183678  -0.8612361  -0.04050999]
Sequence Outputs Shape:(2, 128, 128)
Sequence Outputs Values:[[-0.57619923 -1.4529747  -2.635678   ... -0.78815436 -2.3346112
   0.10347109]
 [-1.1001809  -0.587732    0.44278532 ... -2.1196795  -1.8466102
   0.12794504]
 [-0.88154525 -0.44954467  0.2264989  ... -1.4267368  -1.9974496
   0.08139309]
 ...
 [-0.26786882 -0.8721406   0.7460685  ... -1.3083892  -1.3133664
   1.4343864 ]
 [-0.23225668 -0.5403946  -0.19768439 ... -0.26064742 -1.9160788
   1.5167722 ]
 [ 0.34898573 -0.44532895 -0.19966516 ... -1.0288548  -2.6912522
   0.7270997 ]]


In [ ]:
def build_classifier_model():
  text_input = tf.keras.layers.Input(shape=(), dtype=tf.string, name='text')
  preprocessing_layer = hub.KerasLayer(tfhub_handle_preprocess, name='preprocessing')
  encoder_inputs = preprocessing_layer(text_input)
  encoder = hub.KerasLayer(tfhub_handle_encoder, trainable=True, name='BERT_encoder')
  outputs = encoder(encoder_inputs)
  net = outputs['pooled_output']
  net = tf.keras.layers.Dropout(0.1)(net)
  net = tf.keras.layers.Dense(1, activation=None, name='classifier')(net)
  return tf.keras.Model(text_input, net)

In [ ]:
classifier_model = build_classifier_model()
bert_raw_result = classifier_model(tf.constant(text_test))
print(tf.sigmoid(bert_raw_result))

tf.Tensor(
[[0.71271133]
 [0.8688891 ]], shape=(2, 1), dtype=float32)


In [ ]:
loss = tf.keras.losses.BinaryCrossentropy(from_logits=True)
metrics = tf.metrics.BinaryAccuracy()

In [ ]:
epochs = 3
steps_per_epoch = tf.data.experimental.cardinality(train_ds).numpy()
num_train_steps = steps_per_epoch * epochs
num_warmup_steps = int(0.1*num_train_steps)

init_lr = 3e-5
optimizer = optimization.create_optimizer(init_lr=init_lr,
                                          num_train_steps=num_train_steps,
                                          num_warmup_steps=num_warmup_steps,
                                          optimizer_type='adamw')

In [ ]:
classifier_model.compile(optimizer=optimizer,
                         loss=loss,
                         metrics=metrics)

In [ ]:
print(f'Training model with {tfhub_handle_encoder}')
history = classifier_model.fit(x=train_ds,
                               validation_data=val_ds,
                               epochs=epochs)

Training model with https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-2_H-128_A-2/1
Epoch 1/3
625/625 [==============================] - 214s 337ms/step - loss: 0.6462 - binary_accuracy: 0.6072 - val_loss: 0.5127 - val_binary_accuracy: 0.7382
Epoch 2/3
625/625 [==============================] - 210s 337ms/step - loss: 0.4930 - binary_accuracy: 0.7491 - val_loss: 0.4575 - val_binary_accuracy: 0.7812
Epoch 3/3
625/625 [==============================] - 211s 338ms/step - loss: 0.4463 - binary_accuracy: 0.7832 - val_loss: 0.4484 - val_binary_accuracy: 0.7850


In [ ]:
test_acc2= classifier_model.evaluate(test_ds)

print('accuracy for Bert model is %.03f'%(test_acc2[1]*100))

782/782 [==============================] - 78s 100ms/step - loss: 0.4510 - binary_accuracy: 0.7817
accuracy for Bert model is 78.168


# **Accuracy for Bert model is 78.168**